### ISSUE TITLE
Automatically adds "Thank you"

### ISSUE THREAD

**Issue by tikmaster** [ggml-org/whisper.cpp#1592]
Testing the large v3 model on a word-by-word transcript output, when there is no audio at the end, it always adds "Thank you"

**Comment by bobqianic** [comment]
That's hallucination.

https://arxiv.org/abs/2311.14648

**Comment by gkarmas** [comment]
Interesting thanks for sharing. Is this fixable on the model? I'm stripping it programmatically for now

**Comment by shylock74** [comment]
[https://github.com/openai/whisper/discussions/928](https://github.com/openai/whisper/discussions/928)

**Comment by misutoneko** [comment]
As I've mentioned in that openai whisper thread, I got rid of these with the --suppress_tokens command line switch.
Looks like whisper.cpp doesn't have that, but the BEG token can be suppressed in whisper_process_logits(),
just add this line:
`logits[vocab.token_beg] = -INFINITY;`

It will cause you to get descriptions of sound events instead of "Thank you".
That's a bit easier to deal with, I think.

**Comment by JRWSP** [comment]
> As I've mentioned in that openai whisper thread, I got rid of these with the --suppress_tokens command line switch. Looks like whisper.cpp doesn't have that, but the BEG token can be suppressed in whisper_process_logits(), just add this line: `logits[vocab.token_beg] = -INFINITY;`
> 
> It will cause you to get descriptions of sound events instead of "Thank you". That's a bit easier to deal with, I think.

Can you give more details, where to add the line into? I don't know c++.

**Comment by bobqianic** [comment]
@JRWSP Add it to this function. https://github.com/ggerganov/whisper.cpp/blob/3163090d89c47933d7c2a080b224f0d2e842b468/whisper.cpp#L4535

**Comment by bobqianic** [comment]
> As I've mentioned in that openai whisper thread, I got rid of these with the --suppress_tokens command line switch. Looks like whisper.cpp doesn't have that, but the BEG token can be suppressed in whisper_process_logits(), just add this line: `logits[vocab.token_beg] = -INFINITY;`
> 
> It will cause you to get descriptions of sound events instead of "Thank you". That's a bit easier to deal with, I think.

Could you explain how removing the BEG token (begin time stamps) helps in reducing hallucinations?

**Comment by misutoneko** [comment]
Well if I've understood this correctly, suppressing the non-speech tokens causes the BEG token to emerge somehow, and that's what causes these hallucinations.
Simply suppressing BEG might help, or not.
The workaround that I used for whisper/whisper-timestamped was to allow non-speech tokens.
Here's the original thread:
https://github.com/linto-ai/whisper-timestamped/discussions/107

I suppose this could all be fixed in the training data too, but that's something we plebs don't get to see.
Btw I have tested this mostly with medium and small models (haven't tried large-v3). The "en" models use a different token id.

**Comment by misutoneko** [comment]
Hmmm, do these hallucinated tokens always have low probability?
Because if so, they could be easily filtered out based on that.
But there could be a risk is that some useful tokens might get lost (low-quality audio?).

Another idea I haven't seen mentioned is that prompting can sometimes help (for short clips?).
Even if the prompt is just " " that will change the output.

**Comment by helpforsoftware** [comment]
i tested on large-turbo v3- TR language word timestamps open max_len=1
 turkish wave file added to post
[(http://www.hasenat.net/mp3files/meal/sure/059hasr.mp3]  (converted to wav 16000)
adds at the end
"altyazı  M.K."
(in english it means sth like  "subtitles done by M.K." )

![Image](https://github.com/user-attachments/assets/493607c9-a035-4bef-8e51-07da845997b2)
